File: lib/forum-wrapper.js
/*
 * Copyright (C) 2014 TopCoder Inc., All Rights Reserved.
 *
 * @version 1.1, bugbuka
 * @author freegod
 *
 * changes in 1.1:
 * add init function, if the jboss is restarted, force to init the _wrapper.
 */

"use strict";
/*jslint stupid: true */

var java = require('java');
var _ = require('underscore');
var config = require('./config');
var sys = require('sys');
var EventEmitter = require('events').EventEmitter;

var DEFAULT_FORUM_HOST = 'jnp://env.topcoder.com:1199';

// Class path configuration for Forum Wrapper
java.classpath = java.classpath.concat([__dirname + '/../build/lib/commons-logging-1.1.3.jar',
    __dirname + '/../build/lib/forum-services.jar',
    __dirname + '/../build/lib/forum-wrapper.jar',
    __dirname + '/../build/lib/forums.jar',
    __dirname + '/../build/lib/gson-2.2.4.jar',
    __dirname + '/../build/lib/javax.ejb.jar',
    __dirname + '/../build/lib/jbossall-client.jar',
    __dirname + '/../build/lib/jivebase_modified.jar',
    __dirname + '/../build/lib/jiveforums.jar']);

// Leave signal handling to upper layer. JVM will ignore system singals.
// See http://docs.oracle.com/javase/6/docs/technotes/tools/solaris/java.html
// for more information.
java.options.push('-Xrs');

/**
 * Create a new instance of forum wrapper
 *
 * @param {String} devForumJNDI the software Forum JNDI url
 * @constructor
 * @classdesc
 * This class wraps the <code>Forums</code> and provides methods that are asynchronous
 * with callback parameter.
 */
function ForumWrapper(devForumJNDI) {
    EventEmitter.call(this);
    var self = this,
        forumHost = devForumJNDI || config.forumHost || DEFAULT_FORUM_HOST;
    
    self._forumHost = forumHost;
    self._wrapper = java.callStaticMethodSync('com.topcoder.node.forum.ForumWrapper',
            "getForumsInstance",
            forumHost, false);

    /**
     * init the _wrapper.
     *
     * @param callback
     *          callback function
     */
    self.init = function (callback) {
        try {
            self._wrapper = java.callStaticMethodSync('com.topcoder.node.forum.ForumWrapper',
                "getForumsInstance", self._forumHost, true);
        } catch (ex) {
            callback(ex, self);
            return;
        }
    };
}

sys.inherits(ForumWrapper, EventEmitter);

/**
 * Assign role to user.
 *
 * @param userId
 *          the user id
 * @param jiveGroup
 *          jive group id or jive group name
 * @param callback
 *          callback function
 */
ForumWrapper.prototype.assignRole = function (userId, jiveGroup, callback) {
    var self = this;
    if (typeof callback !== "function") {
        self.emit('error', "Callback must be a function");
        callback(null, self);
        return;
    }
    self._wrapper.assignRole(userId, jiveGroup, function (err) {
        if (err) {

            self.init(callback);
            self._wrapper.assignRole(userId, jiveGroup, function (err) {
                callback(err, self);
            });
        } else {
            callback(null, self);
        }
    });
};

/**
 * Remove role from user.
 *
 * @param userId
 *          the user id
 * @param jiveGroup
 *          jive group id or jive group name
 * @param callback
 *          callback function
 */
ForumWrapper.prototype.removeRole = function (userId, jiveGroup, callback) {
    var self = this;
    if (typeof callback !== "function") {
        self.emit('error', "Callback must be a function");
        callback(null, self);
        return;
    }
    self._wrapper.removeRole(userId, jiveGroup, function (err) {
        if (err) {

            self.init(callback);
            self._wrapper.removeRole(userId, jiveGroup, function (err) {
                callback(err, self);
            });
        } else {
            callback(null, self);
        }
    });
};

ForumWrapper.prototype.removeUserPermission = function (userId, forumCategoryID, callback) {
    var self = this;
    if (typeof callback !== "function") {
        self.emit('error', "Callback must be a function");
        callback(null, self);
        return;
    }
    self._wrapper.removeUserPermission(userId, forumCategoryID, function (err) {
        if (err) {

            self.init(callback);
            self._wrapper.removeUserPermission(userId, forumCategoryID, function (err) {
                callback(err, self);
            });
        } else {
            callback(null, self);
        }
    });
};

ForumWrapper.prototype.deleteCategoryWatch = function (userId, forumCategoryID, callback) {
    var self = this;
    if (typeof callback !== "function") {
        self.emit('error', "Callback must be a function");
        callback(null, self);
        return;
    }

    self._wrapper.deleteCategoryWatch(userId, forumCategoryID, function (err) {
        if (err) {

            self.init(callback);
            self._wrapper.deleteCategoryWatch(userId, forumCategoryID, function (err) {
                callback(err, self);
            });
        } else {
            callback(null, self);
        }
    });
};

module.exports = ForumWrapper;

File: src/com/topcoder/node/forum/ForumWrapper.java
/**
 * Copyright (C) 2014 TopCoder Inc., All Rights Reserved.
 */
package com.topcoder.node.forum;

import com.topcoder.web.ejb.forums.Forums;
import com.topcoder.web.ejb.forums.ForumsHome;

import javax.naming.Context;
import javax.naming.InitialContext;
import java.util.Properties;

/**
 * <p>
 * Wrapper class to communicate with ForumEJB.
 * </p>
 *
 * <p>
 * Version 1.1 - Fix the issue that if jboss is restarted, the ForumWrapper is no longer usable, and the API call will be problem.
 * <ul>
 * <li>Updated {@link #getForumsInstance(String, Boolean)} method, add param forceInit to control whether force to init the forumService</li>
 * </ul>
 * </p>
 *
 * @author freegod, bugbuka
 * @version 1.1
 * @since 1.0
 */
public class ForumWrapper {

    /**
     * Stands for the default forum host.
     */
    private static final String DEFAULT_FORUM_HOST = "jnp://env.topcoder.com:1199";

    /**
     * Stands for the forum ejb retrieved via JNDI.
     */
    private static Forums forumService;

    /**
     * <p>
     *     Retrieve EJB via JNDI and initialize this wrapper.
     * </p>
     * @param forumHost
     *          stands for the forum host
     * @throws Exception
     *          if any error occurs
     */
    private static void init(String forumHost) throws Exception {
        Properties jndiProps = new Properties();
        jndiProps.put(Context.INITIAL_CONTEXT_FACTORY, "org.jnp.interfaces.NamingContextFactory");
        jndiProps.put(Context.URL_PKG_PREFIXES, "org.jboss.naming:org.jnp.interfaces");
        jndiProps.put(Context.PROVIDER_URL, forumHost);
        Context context = new InitialContext(jndiProps);

        ForumsHome forumsHome = (ForumsHome) context.lookup(ForumsHome.EJB_REF_NAME);
        forumService = forumsHome.create();
    }

    /**
     * <p>
     *     Return <code>Forums</code> instance.
     * </p>
     *
     * @param forumHost
     *          stands for the forum host
     * @param forceInit
     *          if true, do init forumService.
     * @return
     *          <code>Forums</code> instance
     * @throws Exception
     *          if any error occurs
     */
    public static Forums getForumsInstance(String forumHost, boolean forceInit) throws Exception {
        if (null == forumHost) {
            forumHost = DEFAULT_FORUM_HOST;
        }
        if (null == forumService || forceInit) {
            init(forumHost);
        }
        return forumService;
    }

}
